### PR TITLE
common/bl, kv, tests: drop MemDB and simplify buffer::ptr and buffer::raw

### PR DESCRIPTION
**Pull request by rzarzynski**
It seems that the `MemDB` implementation of our `KeyValueDB` abstraction is an experimental thing which is no longer maintained nor used (this claim requires verification!). Dropping it would allow to simplify the entire stack of the `buffer` library as `MemDB` was the sole major user of the buffer cloning feature.

TODO: bump the symbols version for `buffer`.

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by tchaikov** [comment]
dropped from my batch, as it conflicts with #36283

**Comment by rzarzynski** [comment]
Rebased and fixed the merge conflict.

File: src/kv/MemDB.cc [hunk 1]
```diff
@@ -1,653 +0,0 @@
-// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
```
Comment (tchaikov) [inline src/kv/MemDB.cc:1]: i guess the idea of MemDB is used as a stub for understanding the performance impact on the i/o path of using a KV store backend like RocksDB. so we can short-circuit this part when we want to analyze the performance characteristics of the whole system assuming rocksdb is not the bottleneck.

Comment (rzarzynski) [reply src/kv/MemDB.cc:1]: Yup, that's my understanding too. However, is `MemDB` actively used?

**Comment by ljflores** [comment]
@rzarzynski any updates?

**Review by tchaikov — APPROVED** [review]
per the none-responses on the mailing list, i am approving this changeset.

**Comment by rzarzynski** [comment]
The recent push has changed only the last commit's title (thinko: `char_raw` -> `raw_char`); there was no change to the code.

**Comment by sseshasa** [comment]
**Teuthology Test Result**
https://pulpito.ceph.com/yuriw-2022-06-27_15:15:17-rados-wip-yuri2-testing-2022-06-24-1331-distro-default-smithi/

**Related Failure**
@rzarzynski  The following failed job appears to be related to this PR. Can you please take a look?
**JobID:** https://pulpito.ceph.com/yuriw-2022-06-27_15:15:17-rados-wip-yuri2-testing-2022-06-24-1331-distro-default-smithi/6901044
**Failure Reason:** Command crashed: "sudo TESTDIR=/home/ubuntu/cephtest bash -c 'mkdir $TESTDIR/kvtest && cd $TESTDIR/kvtest && ceph_test_keyvaluedb'"
**Assertion failure** : In test KeyValueDB/KVTest.OpenClose

**Comment by rzarzynski** [comment]
The QA error was due to a missed removal of MemDB testing. Fixed in https://github.com/ceph/ceph/pull/36282/commits/d105837d6a0f1ac54725a6a1a3577571b9bf21d6. I think the PR is ready for another run.

File: src/blk/kernel/KernelDevice.cc [hunk 2]
```diff
@@ -1065,11 +1065,6 @@ struct ExplicitHugePagePool {
       // don't delete nor unmmap; recycle the region instead
       region_q.push(data);
     }
-    raw* clone_empty() override {
-      // the entire cloning facility is used solely by the dev-only MemDB.
-      // see: https://github.com/ceph/ceph/pull/36282
```
Comment (ronen-fr) [inline src/blk/kernel/KernelDevice.cc:1070]: I'm confused. How can see this comment in the "removed" side of PR#36282?

Comment (ronen-fr) [reply src/blk/kernel/KernelDevice.cc:1070]: Ah... already envisioned in https://github.com/ceph/ceph/pull/43849

**Review by idryomov — APPROVED** [review]
Ack from the RBD perspective

**Comment by tchaikov** [comment]
jenkins test api

**Comment by rzarzynski** [comment]
jenkins test api

**Comment by rzarzynski** [comment]
jenkins test api

**Comment by yuriw** [comment]
jenkins test make check

**Pull request merged**